coding_bestpractices.rst

```diff
@@ -0,0 +1,573 @@
+
+Bad:
+
+.. code:: igor
+
+       Function CreateWave(param)
+           variable param
+
+           Make myWave
+       End
+
+       Function someOtherFunction()
+
+           WAVE/Z myWave
+
+           if(!WaveExists(myWave))
+               CreateWave(param)
+           endif
+
+           WAVE myWave
+       End
+
+Good:
+
+.. code:: igor
+
+       Function/Wave GetWave(param)
+           variable param
+
+           // GetWaveFolder has the same logic as GetWave
+           // Creates the datafolder if it does not exist
+           // and returns a reference to it
+           DFREF dfr = GetWaveFolder(param)
+
+           WAVE/Z/SDFR=dfr data
+           if(WaveExists(data))
+               return data
+           endif
+
+           Make dfr:data/Wave=data
+           // fill wave with default content
+
+           return data
+       End
+
+       Function someOtherFunction()
+           Wave wv = GetWave(param)
+
+           // code
+       End
+
+Reasons:
+
+-  More reliable to use
+
+-  Less code at the call site
+
+-  Avoids code duplication at the call site
+
+-  Allows to handle changes to the wave structure in one place
+   (centralized resource management)
+
+Unnecessary use of StringMatch
+------------------------------
+
+Bad:
+
+.. code:: igor
+
+       if(StringMatch(str, ""))
+           // code
+       endif
+
+       if(StringMatch(str, "abcd"))
+           // code
+       endif
+
+Good:
+
+.. code:: igor
+
+       // can be written with a helper function if(isEmpty(str))
+       if(!cmpstr(str, ""))
+           // ...
+       endif
+
+       // StringMatch is only required if you want to
+       // use ! or * in the second parameter
+       if(!cmpstr(str, "abcd"))
+           // ...
+       endif
+
+Reasons:
+
+-  Faster
+
+-  Better readability
+
+Unnecessary loops
+-----------------
+
+Bad:
+
+.. code:: igor
+
+       for(i = 10; i < 101; i += 1)
+           mywave[i] = i^2
+       endfor
+
+Good:
+
+.. code:: igor
+
+       mywave[10, 100] = p^2
+
+Reasons:
+
+-  Much faster (at least ten times)
+
+See also ``DisplayHelpTopic "Waveform Arithmetic and Assignment"`` for
+an in-depth introduction to the topic.
+
+Unnecessary use of sprintf
+--------------------------
+
+Bad:
+
+.. code:: igor
+
+       string str
+       sprintf str, "%s" GetName()
+
+Good:
+
+.. code:: igor
+
+       string str = GetName()
+
+Reasons:
+
+-  Better readability
+
+-  Shorter
+
+-  Faster
+
+Unnecessary use of Execute
+--------------------------
+
+Bad:
+
+.. code:: igor
+
+       string cmd
+       sprintf cmd, "wv = 1 + 2"
+       Execute cmd
+
+Good:
+
+.. code:: igor
+
+       wv = 1 + 2
+
+Reasons:
+
+-  Better readability
+
+-  Shorter
+
+-  Faster
+
+Avoid relying on the top window and prefer structure based GUI control procedures
+---------------------------------------------------------------------------------
+
+Bad:
+
+.. code:: igor
+
+       Function ButtonProc(ctrlName) : ButtonControl
+           String ctrlName
```

Comment:
`ctrlName` is never used again.

Reply:
This is an IP6 style function argument. I've now switched to IP7 style ones in this file as well.